### PR TITLE
Bk-LSTM

### PR DESCRIPTION
**Pull request by evanyeohboonkhai**
Please check all the file name start with LSTM

### CODE REVIEW

**Review by KianYang-Lee — CHANGES_REQUESTED** [review]
Error in 04 - Univariate Single Step LSTM.ipynb. Please correct.

<code>NameError                                 Traceback (most recent call last)
<ipython-input-29-303829621deb> in <module>
      4 
      5 # Build Scaler and scale the train and test data
----> 6 scaler = MinMaxScaler(feature_range=(-1, 1))
      7 train_data_normalized_shampoo = scaler.fit_transform(train_data_shampoo)
      8 test_data_normalized_shampoo = scaler.fit_transform(test_data_shampoo)

NameError: name 'MinMaxScaler' is not defined</code>

**Comment by evanyeohboonkhai** [comment]
Error in 04 - Univariate Single Step LSTM.ipynb.  solved 
This is a duplicated file, hence I remove it. Thanks @KianYang-Lee

File: notebook/Solution/04 - LSTM -Multivariate Input -Univariate Output-MultiStep.ipynb [hunk 1]
```diff
@@ -0,0 +1,2024 @@
...
+       "5                  207.364  "
+      ]
+     },
+     "execution_count": 29,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "demand_slice = demand.iloc[:,2::]\n",
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - LSTM -Multivariate Input -Univariate Output-MultiStep.ipynb]: 2nd colon kinda redundant if you are not specifying any value? Maybe remove it?

File: notebook/Solution/04 - LSTM -Multivariate Input -Univariate Output-MultiStep.ipynb [hunk 2]
```diff
@@ -0,0 +1,2024 @@
...
+    "testScore_demand = math.sqrt(mean_squared_error(y_test_demand[:,0], y_test_prediction_demand[:,0]))\n",
+    "print('Test Score: %.2f RMSE' % (testScore_demand))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Plot Explanation\n",
+    "- Blue Line-Indicate as test data before sequencing \n",
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - LSTM -Multivariate Input -Univariate Output-MultiStep.ipynb]: Again, believe you can come up with a better explanation. Thanks. Now it is not really intuitive.

**Review by KianYang-Lee — CHANGES_REQUESTED** [review]
Looked at the flow and logic of the codes, did not check the Markdown so will be suggesting based on flow and logic soundness, which were feedback in each file. General comments are:

1. A lot of repetitive parts across notebooks, it would be helpful to put them in a separate module and just call them as methods.
2. Some showcase of basic data analytics would be helpful to prime the viewers on how to perform effective data analysis, and also to prep them for the exam since we are going to check with them on that part.
3. Model performance wise looks like better compared to CNN.

Overall, a job well done. Hope that you can make some modifications based on the input and we will be ready to merge after that basing on the code logic.

**Review by scot-certifai — APPROVED** [review]

**Comment by KianYang-Lee** [comment]
image
Awaiting modifications from branch owner before merging. Avoid merging first to all until confirmation from @evanyeohboonkhai  thanks

**Review by KianYang-Lee — APPROVED** [review]

**Comment by scot-certifai** [comment]
Can this branch be merged already @evanyeohboonkhai ?

**Comment by evanyeohboonkhai** [comment]
> Can this branch be merged already @evanyeohboonkhai ?

not yet, currently, update with the instruction

**Comment by evanyeohboonkhai** [comment]
please have a review in the following LSTM file 
1. 04-LSTM-Univariate-Single-Step.ipynb
2. 04-LSTM-Univariate-Multi-Step.ipynb
3. 04-LSTM-Multivariate-Input-Univariate-Output-Single-Step.ipynb
4. 04-LSTM-Multivariate-Input-Univariate-Output-Multi-Step.ipynb

Add in data_module.py , deep_learning_module.py to modularize the function require 

Remove Multivariate input , multivariate output.ipynb due to need more time research , will include in future

File: notebook/Solution/04 - LSTM-Multivariate-Input-Univariate-Output-Multi-Step.ipynb [hunk 3]
```diff
@@ -0,0 +1,3000 @@
...
...bopc8rgsMuhaKoJYGrWm8Bj06vm+aQXwuyzTfCVlmdea23ea/fH8PglJtA+4BwTqLTWmmxy3SazzOKWN01gmFli/lD4/K/NcIQlvzb7pp1mgkyAT58wGdRpp5mAIJqtb8Gj50aywWD+e087NRI4Nfcy4cyWAjO/ZLar/h37uSaL3RnfH672VPOcNr3a9Xn0xpFuglq/x2TQg/5Ir0E8HOnm96nq0/iv6Y3NYQLEvo5D5JyyQ03g3bTT/LwfeJ7pzVh6OxRNMz93G1+Gz//K/EGwY1mk12DHMjOWedwRcPCFUDw9dO+guZ+vzfyBAmZs+Fu/gwufMvcXQiTFvhyUXg7cA1ymtX6g03mLgEUA48aNO2T79u1D08BNr5kxTcNFapb58rWbzE8ys7hZo0wQpYPmKxgw2SNXtsnu2Oxmf255zwAs4DcZP39H1y5jse9pqzc/r6lZJmtbNN2soGMJ+Mzs/rZakzVMyzMZNutnRmsTfNscJoPorjLZR18bdLjNz1lWqcmwed0m2A14Qz+XOvLzmZJqssltdaHehjQTUFq/R64c0xZlM+d2NJvMYFqe+X9Aa615r9SsULuCZonKUQeGhnso0466TSarmjPOfAZ3jfn5ziwxvxutdSZrnVFoup79HkCbYR4pqaHfIwUTjg197zdbMBnh+s3m+bTVmWeQlm96JPwe83tls5s/qvImmN6I2o2RjLq31WRndQAqPzWZ1RSX+T3OHWf++GlvMPdMyzO/nylOKJljfl+bdpnnrwOmndYfPGPmmaDSXWU+S/aYofud7u9YaSHEoNuXglKr+/7HWuvfDbvueyGEEEII0au+gtK4Zt8PIy8A1cCVSqkW4FLMxKclSWyTEEIIIYTYS7GK5w8rWmsPcA7gBu7EBKjn9DbzXgghhBBCjAwjLVOK1votIMFFGoUQQgghxFAaUZlSIYQQQgixbxpRE50GQilVAwzR9PuEygGakvTehUBtkt47lmQ+l72R6Gc6Up/L3oj1TPfHZxKPWM9lOP/+J0qif1ZG6jMdzr9DyXqmw/mZDMRgPce+nst4rXVRtAP7fFC6r1BK3ae1XpSk917e20y5ZEvmc9kbiX6mI/W57I1Yz3R/fCbxiPVchvPvf6Ik+mdlpD7T4fw7lKxnOpyfyUAM1nMc6HOR7vuR47lkN2CYkucSnTyXnuSZRCfPpSd5JtHJc
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - LSTM-Multivariate-Input-Univariate-Output-Multi-Step.ipynb:2963]: I think this a mistake of strikethrough the other notebooks?

File: notebook/Solution/04 - LSTM-Univariate-Multi-Step.ipynb [hunk 4]
```diff
@@ -0,0 +1,2345 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 47,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "### BEGIN SOLUTION\n",
+    "train_iter_shampoo , test_iter_shampoo = iterator(train_shampoo_dict, test_shampoo_dict)\n",
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - LSTM-Univariate-Multi-Step.ipynb:1832]: TypeError, please resolve
image

File: notebook/Solution/04 - LSTM-Univariate-Single-Step.ipynb [hunk 5]
```diff
@@ -0,0 +1,3430 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Data Iterator\n",
+    "Iterator is created to allow the data separate into several batches to fasten the training process\n",
+    ">**Instruction** : <br>\n",
+    "    1. First, use `TensorDataset` to \"wrap\" the x-feature and y-label together.<br>\n",
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - LSTM-Univariate-Single-Step.ipynb]: People either refer to feature as "feature" or "x", and label as "label" or "y"

File: notebook/Solution/04 - LSTM-Univariate-Single-Step.ipynb [hunk 6]
```diff
@@ -0,0 +1,3430 @@
...
+    "\n",
+    "single_step_plot(original_test_data = test_data,\n",
+    "                 sequence_test_data = y_test,\n",
+    "                 forecast_data = y_test_prediction,\n",
+    "                 test_time = test_time,\n",
+    "                 window_size = window_size,\n",
+    "                 original_plot = True)   "
+   ]
+  },
+  {
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - LSTM-Univariate-Single-Step.ipynb]: Pelase remove empty cell

File: .gitignore [hunk 7]
```diff
@@ -2,4 +2,8 @@
 .ipynb_checkpoints
 *.iml
 
-target
\ No newline at end of file
+target
+*.pyc
+*.pyc
```
Comment (KianYang-Lee) [inline .gitignore:7]: Can remove redundant lines

File: notebook/Solution/data_module.py [hunk 8]
```diff
@@ -0,0 +1,388 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[3]:
```
Comment (KianYang-Lee) [inline notebook/Solution/data_module.py]: Remove In[3] and other redundant lines

File: notebook/Solution/04 - LSTM-Multivariate-Input-Univariate-Output-Single-Step.ipynb [hunk 9]
```diff
@@ -0,0 +1,3205 @@
...
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "# pip install this package to view the summary of LSTM due to original torchsummary does have the function to view the LSTM\n",
+    "# used jupyter install due to it does not have conda version\n",
+    "# %%capture suppress information of torchsummaryX installation\n",
+    "!pip install torchsummaryX"
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - LSTM-Multivariate-Input-Univariate-Output-Single-Step.ipynb:64]: suggest to include torchsummaryX to environment

**Review by KianYang-Lee — APPROVED** [review]
Great job, made some suggestions to further improve the slide

File: notebook/Solution/04 - LSTM-Multivariate-Input-Univariate-Output-Single-Step.ipynb (hunk 9 above)
Comment (evanyeohboonkhai) [reply notebook/Solution/04 - LSTM-Multivariate-Input-Univariate-Output-Single-Step.ipynb]: image
cannot found torchsummaryX in conda environment, hence use pip install

**Review by KianYang-Lee — APPROVED** [review]
Already reviewed and approved

**Comment by evanyeohboonkhai** [comment]
@KianYang-Lee  Please merge if it's ok for you

**Review by scot-certifai — APPROVED** [review]

**Pull request merged**